### DISCUSSION TITLE
AGC/AGE Algorithm

### DISCUSSION THREAD

**Discussion by dgalland** [raspberrypi/picamera2#795] [General]
I wonder about the convergence of the AGC/AEC algorithm.
Below is the sequence of queries in the following case:
Start with auto exposure 7328, then a fixed exposure of twice this value 14657 and return to Auto.
The AGC/AEC algorithm seems to take a long time to converge, more than 20 requests.
There seem to be two or three steps in the behavior
Can't rely on AeLocked to know when the algorithm has converged
Note: all this with the tuning file and all default parameters (I note that 10000 is a limit value in exposure_modes)

Returning to auto
AeLocked: False Exposure: 14657  Analog: 1.0  Digital: 1.001832365989685
AeLocked: False Exposure: 14657  Analog: 1.0  Digital: 1.001832365989685
AeLocked: True Exposure: 14657  Analog: 1.0  Digital: 1.001832365989685
AeLocked: False Exposure: 14657  Analog: 1.0  Digital: 1.001832365989685
AeLocked: False Exposure: 14657  Analog: 1.0  Digital: 1.001832365989685
AeLocked: False Exposure: 14657  Analog: 1.0  Digital: 1.001832365989685
AeLocked: False Exposure: 14657  Analog: 1.0  Digital: 1.001832365989685
AeLocked: False Exposure: 10000  Analog: 1.3350716829299927  Digital: 1.0002212524414062
AeLocked: False Exposure: 10000  Analog: 1.2278177738189697  Digital: 1.000914454460144
AeLocked: False Exposure: 10000  Analog: 1.1428571939468384  Digital: 1.0008269548416138
AeLocked: False Exposure: 10000  Analog: 1.0756303071975708  Digital: 1.0000661611557007
AeLocked: False Exposure: 10000  Analog: 1.020937204360962  Digital: 1.0002498626708984
AeLocked: True Exposure: 9742  Analog: 1.0  Digital: 1.0010063648223877
AeLocked: True Exposure: 9342  Analog: 1.0  Digital: 1.002004861831665
AeLocked: True Exposure: 9028  Analog: 1.0  Digital: 1.0002583265304565
AeLocked: True Exposure: 8400  Analog: 1.0  Digital: 1.0010716915130615
AeLocked: True Exposure: 8028  Analog: 1.0  Digital: 1.0014674663543701
AeLocked: True Exposure: 7800  Analog: 1.0  Digital: 1.0021942853927612
AeLocked: True Exposure: 7657  Analog: 1.0  Digital: 1.002753496170044
AeLocked: True Exposure: 7571  Analog: 1.0  Digital: 1.002395749092102
AeLocked: True Exposure: 7514  Analog: 1.0  Digital: 1.000394582748413
AeLocked: True Exposure: 7457  Analog: 1.0  Digital: 1.0008976459503174
AeLocked: True Exposure: 7400  Analog: 1.0  Digital: 1.0035778284072876
AeLocked: True Exposure: 7400  Analog: 1.0  Digital: 1.0001429319381714
AeLocked: True Exposure: 7371  Analog: 1.0  Digital: 1.0017125606536865
AeLocked: True Exposure: 7371  Analog: 1.0  Digital: 1.0002228021621704
AeLocked: True Exposure: 7342  Analog: 1.0  Digital: 1.003018856048584

**Comment by davidplowman** [comment]
Hi, yes I agree that the AeLocked metadata needs to be revised. Possibly it should also leap more quickly to its final target value once it's very close. There's always lots to think about!

**Reply by dgalland** [reply]
Hi, I'm coming back to this discussion because there hasn't been much change in the AGC/AE algorithm.

I do the following tests (full resolution,  still configuration 2 buffers imx477.json )

1. Start the camera, wait 2s then switch to a fixed exposure

> START AeLocked: False Exposure: 5600  Analog: 1.0  Digital: 1.0189465284347534
> Going to fixed 1000
> AeLocked: False Exposure: 5600  Analog: 1.0  Digital: 1.0189465284347534
> AeLocked: False Exposure: 5685  Analog: 1.0  Digital: 1.0035854578018188
> AeLocked: False Exposure: 5685  Analog: 1.0  Digital: 1.0035854578018188
> AeLocked: False Exposure: 5685  Analog: 1.0  Digital: 1.0035854578018188
> AeLocked: False Exposure: 5685  Analog: 1.0  Digital: 1.0036529302597046
> AeLocked: False Exposure: 5685  Analog: 1.0  Digital: 1.0041638612747192
> AeLocked: False Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: False Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: False Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: False Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: False Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: True Exposure: 1000  Analog: 1.0  Digital: 1.0
> ...
> 
> 

You have to wait 7 frames to get a fixed exposure
The behavior of the AeLocked boolean is not consistent
Even if I understand that a gap between a setting and the taking into account of its result, I find that 7 frames is still a lot since we are at 10fps?

2. Return to auto

> Returning to auto
> AeLocked: True Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: True Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: False Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: False Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: False Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: False Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: False Exposure: 1942  Analog: 1.0  Digital: 1.00248122215271
> AeLocked: False Exposure: 2685  Analog: 1.0  Digital: 1.0074456930160522
> AeLocked: False Exposure: 3285  Analog: 1.0  Digital: 1.0080310106277466
> AeLocked: False Exposure: 3771  Analog: 1.0  Digital: 1.0068279504776
> AeLocked: False Exposure: 4171  Analog: 1.0  Digital: 1.0024405717849731
> AeLocked: False Exposure: 4457  Analog: 1.0  Digital: 1.0067144632339478
> AeLocked: False Exposure: 4714  Analog: 1.0  Digital: 1.0033462047576904
> AeLocked: True Exposure: 5142  Analog: 1.0  Digital: 1.0038419961929321
> AeLocked: True Exposure: 5371  Analog: 1.0  Digital: 1.0051066875457764
> AeLocked: True Exposure: 5514  Analog: 1.0  Digital: 1.0025951862335205
> AeLocked: True Exposure: 5571  Analog: 1.0  Digital: 1.0050638914108276
> AeLocked: True Exposure: 5571  Analog: 1.0  Digital: 1.0050638914108276
> ...
> 

Another 7 frames before the AGC/AE algorithm starts. Then It takes about ten frames to converge
The behavior of the AeLocked boolean is not consistent

3. Change speed=1
algo["channels"][0]["speed"] = 1.

> Returning to auto
> AeLocked: True Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: True Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: False Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: False Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: False Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: False Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: False Exposure: 5714  Analog: 1.0  Digital: 1.0042256116867065
> AeLocked: False Exposure: 5714  Analog: 1.0  Digital: 1.0042256116867065
> AeLocked: False Exposure: 5714  Analog: 1.0  Digital: 1.0042256116867065
> AeLocked: False Exposure: 5714  Analog: 1.0  Digital: 1.0042256116867065
> AeLocked: False Exposure: 5714  Analog: 1.0  Digital: 1.0042256116867065
> AeLocked: True Exposure: 5714  Analog: 1.0  Digital: 1.0042256116867065
> AeLocked: True Exposure: 5714  Analog: 1.0  Digital: 1.0042256116867065
> AeLocked: True Exposure: 5714  Analog: 1.0  Digital: 1.0042256116867065
> 

The AGC/AE algorithm immediately finds an acceptable value !!

**Reply by davidplowman** [reply]
You're right. I've not had any time to do any work on it.

As regards the long delay for settings to take effect, that's mostly down to libcamera. Control changes get stuck at the back of the request queue, and even once they get to the front there are still another 3 or frames or so of delay. We've long been wanting to change this, but now, libcamera is reviewing how its API works which should give us a route to improving this. But it will still be a while.

Yes, the AeLocked indicator needs work. There's other stuff that needs doing first, but I will try to get round to it at some point.

**Reply by dgalland** [reply]
To complete I did some tests for the HDR algorithm (on a PI5). The HDR algorithm is a special operating mode of the AGC/AE algorithm (and not an HDR at the sensor level). In the configuration file we define two AGC channels short and long. The first remark is that these channels are very little different in the imx477.json file provided, to obtain a real effect you have to change the parameters, I did it for example with base_ev respectively 0.5 and 1.5 for channels 1 and 2

As before start in fixed 1000 then AeEnable in HDR mode:

> Auto AeLocked: True Exposure: 5571  Analog: 1.0  Digital: 1.0021299123764038  Lux: 1136.2918701171875
> AeLocked: True Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: True Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: True Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: True Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: True Exposure: 1000  Analog: 1.0  Digital: 1.0
> AeLocked: True Exposure: 1000  Analog: 1.0  Digital: 1.0
> HdrMode: 1 HdrChannel: 1 AeLocked: False Exposure: 2257  Analog: 1.0  Digital: 1.0050550699234009
> HdrMode: 1 HdrChannel: 3 AeLocked: False Exposure: 8885  Analog: 1.0  Digital: 1.0000739097595215
> HdrMode: 1 HdrChannel: 1 AeLocked: False Exposure: 2257  Analog: 1.0  Digital: 1.0050745010375977
> HdrMode: 1 HdrChannel: 3 AeLocked: False Exposure: 8857  Analog: 1.0  Digital: 1.0031594038009644
> HdrMode: 1 HdrChannel: 1 AeLocked: False Exposure: 2257  Analog: 1.0  Digital: 1.0024116039276123
> HdrMode: 1 HdrChannel: 3 AeLocked: False Exposure: 8857  Analog: 1.0  Digital: 1.0020825862884521
> HdrMode: 1 HdrChannel: 1 AeLocked: False Exposure: 2257  Analog: 1.0  Digital: 1.0023448467254639
> HdrMode: 1 HdrChannel: 3 AeLocked: False Exposure: 8857  Analog: 1.0  Digital: 1.001430869102478
> HdrMode: 1 HdrChannel: 1 AeLocked: False Exposure: 2257  Analog: 1.0  Digital: 1.0025267601013184
> HdrMode: 1 HdrChannel: 3 AeLocked: False Exposure: 8857  Analog: 1.0  Digital: 1.0013855695724487
> HdrMode: 1 HdrChannel: 1 AeLocked: True Exposure: 2257  Analog: 1.0  Digital: 1.0026646852493286

After the usual 7 frames the algorithm stabilizes immediately (but AeLocked still inconsistent and error on the channel why 3 is reported and not 2?). In unmerged mode we get alternately the short and long frames (on the main and raw streams) and in merge mode a merge frame (on the main stream only).
Below the results from left to right
Auto exposure on channel 0 exposure 5571
Unmerged channels 1 exposure 2257
Unmerged Channel 2 exposure 8857
Merged Channels 1 and 2
The result of the merge is spectacular but it is a "black box" the treatments are poorly documented

![hdr50](https://github.com/user-attachments/assets/289449b5-a9f1-4044-baf7-db97c75cdb77)
